### ISSUE TITLE
CI ROS tests too hard to reproduce

### ISSUE THREAD

**Issue by julianoes** [PX4/PX4-Autopilot#11280]
TL;DR: We need more cohesive docs and scripts to run the CI tests.

To investigate #11257 I set out to reproduce the CI tests on my local machine. Running it in docker was out of the question because it makes catching hardfaults even harder (I tried that unsuccessfully last time).

Therefore, fresh VirtualBox with Ubuntu 16.04, and then I try to follow the instructions as well as possible.

```
wget https://raw.githubusercontent.com/PX4/Devguide/master/build_scripts/ubuntu_sim_nuttx.sh
chmod +x ubuntu_sim_nuttx.sh
./ubuntu_sim_nuttx.sh
```

So far so good, then I needed the ROS dependencies:

```
wget https://raw.githubusercontent.com/PX4/Devguide/master/build_scripts/ubuntu_sim_ros_gazebo.sh
chmod +x ubuntu_sim_ros_gazebo.sh
./ubuntu_sim_ros_gazebo.sh
```

Here I ran into troubles because of Gazebo 9 being installed by `ubuntu_sim_nuttx.sh` but `ubuntu_sim_ros_gazebo.sh` needs Gazebo 7 (see separate [issue](https://github.com/PX4/Devguide/issues/722). I had to remove Gazebo 9, and then run it again.

```
sudo apt-get remove gazebo7 libgazebo7-dev
./ubuntu_sim_ros_gazebo.sh
```
Now, I should be able to run the tests, confusingly the clone is already here, that's magic:

```
cd src/Firmware
git submodule update --init --recursive
test/rostest_px4_run.sh mavros_posix_test_mission.test mission:=FW_mission_1 vehicle:=plane
```

Ok, looks like MAVROS is missing somehow:
```
GAZEBO_PLUGIN_PATH :/home/julianoes/src/Firmware/test/../build/px4_sitl_default/build_gazebo
GAZEBO_MODEL_PATH :/home/julianoes/src/Firmware/test/../Tools/sitl_gazebo/models
LD_LIBRARY_PATH /opt/ros/kinetic/lib:/opt/ros/kinetic/lib/x86_64-linux-gnu:/home/julianoes/src/Firmware/test/../build/px4_sitl_default/build_gazebo
... logging to /home/julianoes/.ros/ros_logs/rostest-ubuntu1604-vbox-6442.log
Traceback (most recent call last):
  File "/opt/ros/kinetic/bin/rostest", line 36, in <module>
    rostestmain()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/__init__.py", line 274, in rostestmain
    _main()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/rostest_main.py", line 159, in rostestmain
    testCase = rostest.runner.createUnitTest(pkg, test_file, options.reuse_master, options.clear, options.results_base_dir)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/runner.py", line 229, in createUnitTest
    config = roslaunch.parent.load_config_default([test_file], None)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/config.py", line 451, in load_config_default
    loader.load(f, config, verbose=verbose)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 749, in load
    self._load_launch(launch, ros_config, is_core=core, filename=filename, argv=argv, verbose=verbose)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 721, in _load_launch
    self._recurse_load(ros_config, launch.childNodes, self.root_context, None, is_core, verbose)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 685, in _recurse_load
    val = self._include_tag(tag, context, ros_config, default_machine, is_core, verbose)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 95, in call
    return f(*args, **kwds)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 627, in _include_tag
    default_machine, is_core, verbose)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 685, in _recurse_load
    val = self._include_tag(tag, context, ros_config, default_machine, is_core, verbose)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 95, in call
    return f(*args, **kwds)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 589, in _include_tag
    inc_filename = self.resolve_args(tag.attributes['file'].value, context)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/xmlloader.py", line 183, in resolve_args
    return substitution_args.resolve_args(args, context=context.resolve_dict, resolve_anon=self.resolve_anon)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/substitution_args.py", line 370, in resolve_args
    resolved = _resolve_args(resolved, context, resolve_anon, commands)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/substitution_args.py", line 383, in _resolve_args
    resolved = commands[command](resolved, a, args, context)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/substitution_args.py", line 151, in _find
    source_path_to_packages=source_path_to_packages)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/substitution_args.py", line 197, in _find_executable
    full_path = _get_executable_path(rp.get_path(args[0]), path)
  File "/usr/lib/python2.7/dist-packages/rospkg/rospack.py", line 203, in get_path
    raise ResourceNotFound(name, ros_paths=self._ros_paths)
rospkg.common.ResourceNotFound: mavros
ROS path [0]=/opt/ros/kinetic/share/ros
ROS path [1]=/opt/ros/kinetic/share
ROS path [2]=/home/julianoes/src/Firmware/test/..
ROS path [3]=/home/julianoes/src/Firmware/test/../Tools/sitl_gazebo
```

So let's add that:
```
sudo apt-get install ros-kinetic-mavros ros-kinetic-mavros-extras
```

And try again:
```
julianoes@ubuntu1604-vbox:~/src/Firmware$ test/rostest_px4_run.sh mavros_posix_test_mission.test mission:=FW_mission_1 vehicle:=plane
GAZEBO_PLUGIN_PATH :/home/julianoes/src/Firmware/test/../build/px4_sitl_default/build_gazebo
GAZEBO_MODEL_PATH :/home/julianoes/src/Firmware/test/../Tools/sitl_gazebo/models
LD_LIBRARY_PATH /opt/ros/kinetic/lib:/opt/ros/kinetic/lib/x86_64-linux-gnu:/home/julianoes/src/Firmware/test/../build/px4_sitl_default/build_gazebo
... logging to /home/julianoes/.ros/ros_logs/rostest-ubuntu1604-vbox-7295.log
[ROSUNIT] Outputting test results to /home/julianoes/.ros/test_results/px4/rostest-test_mavros_posix_test_mission.xml
[ INFO] [1548226559.791537371]: FCU URL: udp://:14540@localhost:14557
Traceback (most recent call last):
  File "/opt/ros/kinetic/lib/gazebo_ros/spawn_model", line 21, in <module>
    import rospy, sys, os, time
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/__init__.py", line 49, in <module>
[ INFO] [1548226559.793863639]: udp0: Bind address: 0.0.0.0:14540
    from .client import spin, myargv, init_node, \
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/client.py", line 52, in <module>
    [ INFO] [1548226559.793957463]: udp0: Remote address: 127.0.0.1:14557
import roslib
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslib/__init__.py", line 50, in <module>
    from roslib.launcher import load_manifest
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslib/launcher.py", line 42, in <module>
    import rospkg
  File "/usr/lib/python2.7/dist-packages/rospkg/__init__.py", line 42, in <module>
    from .rospack import expand_to_packages, get_package_name, \
  File "/usr/lib/python2.7/dist-packages/rospkg/rospack.py", line 35, in <module>
    from xml.etree.cElementTree import ElementTree
  File "/usr/lib/python2.7/xml/etree/cElementTree.py", line 3, in <module>
    from _elementtree import *
  File "<string>", line 22, in <module>
  File "<string>", line 22, in ElementTree
KeyboardInterrupt
[Testcase: testFW_mission_1] ... FAILURE!
FAILURE: Test Fixture Nodes ['sitl'] failed to launch
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/runner.py", line 121, in fn
    self.assert_(not failed, "Test Fixture Nodes %s failed to launch"%failed)
  File "/usr/lib/python2.7/unittest/case.py", line 422, in assertTrue
    raise self.failureException(msg)
--------------------------------------------------------------------------------

[ROSTEST]-----------------------------------------------------------------------

[testFW_mission_1][failed]

SUMMARY
 * RESULT: FAIL
 * TESTS: 0
 * ERRORS: 0
 * FAILURES: 1

ERROR: The following tests failed to run:
 * testFW_mission_1

rostest log file is in /home/julianoes/.ros/ros_logs/rostest-ubuntu1604-vbox-7295.log
```

I'm making an educated guess, maybe I need to build stuff first?
```
DONT_RUN=1 make px4_sitl gazebo
```

Ok, let's try again:
```
GAZEBO_PLUGIN_PATH :/home/julianoes/src/Firmware/test/../build/px4_sitl_default/build_gazebo
GAZEBO_MODEL_PATH :/home/julianoes/src/Firmware/test/../Tools/sitl_gazebo/models
LD_LIBRARY_PATH /opt/ros/kinetic/lib:/opt/ros/kinetic/lib/x86_64-linux-gnu:/home/julianoes/src/Firmware/test/../build/px4_sitl_default/build_gazebo
... logging to /home/julianoes/.ros/ros_logs/rostest-ubuntu1604-vbox-12674.log
[ROSUNIT] Outputting test results to /home/julianoes/.ros/test_results/px4/rostest-test_mavros_posix_test_mission.xml
INFO  [px4] Creating symlink /home/julianoes/src/Firmware/ROMFS/px4fmu_common -> /home/julianoes/.ros/etc
0 WARNING: setRealtimeSched failed (not run as root?)

______  __   __    ___ 
| ___ \ \ \ / /   /   |
| |_/ /  \ V /   / /| |
|  __/   /   \  / /_| |
| |     / /^\ \ \___  |
\_|     \/   \/     |_/

px4 starting.

INFO  [px4] Calling startup script: /bin/sh etc/init.d-posix/rcS 0
  SYS_AUTOSTART: curr: 0 -> new: 1030
...
* NAV_ACC_RAD: curr: 10.0000 -> new: 15.0000
Traceback (most recent call last):
  File "/home/julianoes/src/Firmware/integrationtests/python_src/px4_it/mavros/mission_test.py", line 50, in <module>
    import px4tools
ImportError: No module named px4tools
  RWTO_TKOFF: curr: 0 -> new: 1
  WEST_EN: curr: 0 -> new: 1
[ INFO] [1548227341.408520415]: Plugin global_position initialized
...
[ INFO] [1548227341.654000434]: Plugin param loaded
ERROR [px4] Startup script returned with return value: 2
[ERROR] [1548227341.662487305]: [registerService] Failed to contact master at [ubuntu1604-vbox:41833].  Retrying...
Traceback (most recent call last):
  File "/opt/ros/kinetic/lib/gazebo_ros/spawn_model", line 313, in <module>
    sm.callSpawnService()
  File "/opt/ros/kinetic/lib/gazebo_ros/spawn_model", line 206, in callSpawnService
    rospy.init_node('spawn_model')
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/client.py", line 323, in init_node
    raise rospy.exceptions.ROSInitException("init_node interrupted before it could complete")
rospy.exceptions.ROSInitException: init_node interrupted before it could complete
[Testcase: testFW_mission_1] ... FAILURE!
FAILURE: test [FW_mission_1] did not generate test results
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/runner.py", line 164, in fn
    self.assert_(os.path.isfile(test_file), "test [%s] did not generate test results"%test_name)
  File "/usr/lib/python2.7/unittest/case.py", line 422, in assertTrue
    raise self.failureException(msg)
--------------------------------------------------------------------------------

[ROSTEST]-----------------------------------------------------------------------

[testFW_mission_1][failed]

SUMMARY
 * RESULT: FAIL
 * TESTS: 0
 * ERRORS: 0
 * FAILURES: 1

ERROR: The following tests failed to run:
 * testFW_mission_1

rostest log file is in /home/julianoes/.ros/ros_logs/rostest-ubuntu1604-vbox-12674.log
```

Ok, that's better but now px4tools is missing, that's odd.

```
pip install --user px4tools
```

Let's try again:
```
GAZEBO_PLUGIN_PATH :/home/julianoes/src/Firmware/test/../build/px4_sitl_default/build_gazebo
GAZEBO_MODEL_PATH :/home/julianoes/src/Firmware/test/../Tools/sitl_gazebo/models
LD_LIBRARY_PATH /opt/ros/kinetic/lib:/opt/ros/kinetic/lib/x86_64-linux-gnu:/home/julianoes/src/Firmware/test/../build/px4_sitl_default/build_gazebo
... logging to /home/julianoes/.ros/ros_logs/rostest-ubuntu1604-vbox-14415.log
[ROSUNIT] Outputting test results to /home/julianoes/.ros/test_results/px4/rostest-test_mavros_posix_test_mission.xml
INFO  [px4] Creating symlink /home/julianoes/src/Firmware/ROMFS/px4fmu_common -> /home/julianoes/.ros/etc
0 WARNING: setRealtimeSched failed (not run as root?)

______  __   __    ___ 
| ___ \ \ \ / /   /   |
| |_/ /  \ V /   / /| |
|  __/   /   \  / /_| |
| |     / /^\ \ \___  |
\_|     \/   \/     |_/

px4 starting.

INFO  [px4] Calling startup script: /bin/sh etc/init.d-posix/rcS 0
  SYS_AUTOSTART: curr: 0 -> new: 1030
  ...
[ INFO] [1548227426.728551983]: Plugin vision_pose_estimate loaded
[INFO] [1548227426.752577, 0.000000]: Loading model XML from file
[INFO] [1548227426.753267, 0.000000]: Waiting for service /gazebo/spawn_sdf_model
[ INFO] [1548227426.770889409]: Plugin vision_pose_estimate initialized
[ INFO] [1548227426.771235792]: Plugin vision_speed_estimate loaded
[ INFO] [1548227426.782935678]: Plugin vision_speed_estimate initialized
[ INFO] [1548227426.783261115]: Plugin waypoint loaded
[ INFO] [1548227426.792324926]: Plugin waypoint initialized
[ INFO] [1548227426.792644186]: Plugin wind_estimation loaded
[ INFO] [1548227426.795047724]: Plugin wind_estimation initialized
[ INFO] [1548227426.795303360]: Autostarting mavlink via USB on PX4
[ INFO] [1548227426.795559215]: Built-in SIMD instructions: SSE, SSE2
[ INFO] [1548227426.795632097]: Built-in MAVLink package version: 2018.12.12
[ INFO] [1548227426.795700289]: Known MAVLink dialects: common ardupilotmega ASLUAV autoquad icarous matrixpilot paparazzi slugs standard uAvionix ualberta
[ INFO] [1548227426.795887203]: MAVROS started. MY ID 1.240, TARGET ID 1.1
[INFO] [1548227427.360610, 0.004000]: Calling service /gazebo/spawn_sdf_model
[ INFO] [1548227427.381234468, 0.024000000]: waitForService: Service [/gazebo/set_physics_properties] is now available.
[INFO] [1548227427.614622, 0.184000]: Spawn status: SpawnModel: Successfully spawned entity
[ INFO] [1548227427.641899045, 0.184000000]: Physics dynamic reconfigure ready.
Traceback (most recent call last):
  File "/home/julianoes/src/Firmware/integrationtests/python_src/px4_it/mavros/mission_test.py", line 52, in <module>
    from mavros import mavlink
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/mavros/mavlink.py", line 12, in <module>
    from pymavlink import mavutil
ImportError: No module named pymavlink
INFO  [simulator] Simulator connected on TCP port 4560.
10.10
ERROR [px4] Startup script returned with return value: 2
pure virtual method called
terminate called without an active exception
```

Ok, so pymavlink is also missing.

```
pip install --user pymavlink
```

And we're flying, finally!

```
test/rostest_px4_run.sh mavros_posix_test_mission.test mission:=FW_mission_1 vehicle:=plane
...
```

**Comment by lamping7** [comment]
I know you know this is related, but want to add it here: https://github.com/PX4/Firmware/issues/11024
Also: https://github.com/PX4/containers/pull/107

Normally I use the containers that CI uses to rule out setup differences.

I also noticed that the script you mentioned installs mavros from source, which is a topic that could be discussed.

If we get containers and Jenkins to use these, we can actually maintain these and know when they break.

**Comment by julianoes** [comment]
I'm fighting with this again, this time using docker where `rostest` seems to be missing.

**Comment by julianoes** [comment]
Old, closing.
